File: payola/s2js/runtime/src/main/scala/s2js/runtime/scala/collection/Seq.scala
package s2js.runtime.scala.collection

import s2js.compiler.NativeJs

trait Seq extends Iterable
{
    @NativeJs("[]")
    var internalJsArray = null

    @NativeJs("""
        for (var i in self.internalJsArray) {
            f(self.internalJsArray[i]);
        }
    """)
    def foreach[U](f: Double => U) {}

    @NativeJs("self.internalJsArray.push(x);")
    def +=(x: Any) {}

    // From TraversableLike
    def reversed: Iterable = {
        val elems: Iterable = newInstance
        for (x <- this) {
            elems.prepend(x)
        }
        elems
    }

    @NativeJs("return self.internalJsArray.length;")
    override def size: Int = 0

    @NativeJs("""
        if (s2js.isUndefined(self.internalJsArray[n])) {
            throw new scala.NoSuchElementException('An item with index ' + n + ' is not present.');
        }
        return self.internalJsArray[n];
    """)
    def apply(n: Int): Any = null

    @NativeJs("""
        if (self.size() <= n) {
            throw new scala.NoSuchElementException('An item with index ' + n + ' is not present.');
        }
        self.internalJsArray[n] = newelem;
    """)
    def update(n: Int, newelem: Any) {}

    def length: Int = size

    @NativeJs("""
        if (index < 0 || self.size() <= index) {
            throw new scala.NoSuchElementException('An item with index ' + n + ' is not present.');
        }
        self.internalJsArray.splice(index, 1);
    """)
    def remove(index: Int) {}

    @NativeJs("""self.internalJsArray.splice(0, 0, x);""")
    def prepend(x: Any) {}

    @NativeJs("""
        var index = self.internalJsArray.indexOf(x);
        if (index != -1) {
            self.internalJsArray.splice(index, 1);
        }
    """)
    def -=(x: Double) {}
}

File: payola/web/server/public/javascripts/runtime/s2js/runtime/scala/collection/Seq.js
goog.provide('scala.collection.Seq');
goog.require('scala.collection.Iterable');
goog.require('scala.util.control.Breaks');
scala.collection.Seq = function() {
var self = this;
self.internalJsArray = [];
};
goog.inherits(scala.collection.Seq, scala.collection.Iterable);
scala.collection.Seq.prototype.getInternalJsArray = function() {
var self = this;
return self.internalJsArray;
};
scala.collection.Seq.prototype.setInternalJsArray = function(value) {
var self = this;
self.internalJsArray = value;
};
scala.collection.Seq.prototype.foreach = function(f) {
var self = this;

        for (var i in self.getInternalJsArray()) {
            f(self.getInternalJsArray()[i]);
        }
    };
scala.collection.Seq.prototype.$plus$eq = function(x) {
var self = this;
self.getInternalJsArray().push(x);};
scala.collection.Seq.prototype.reversed = function() {
var self = this;
var elems = self.newInstance();
self.foreach(function(x) {
elems.prepend(x);
});
return elems;

};
scala.collection.Seq.prototype.size = function() {
var self = this;
return self.getInternalJsArray().length;};
scala.collection.Seq.prototype.$apply = function(n) {
var self = this;

        if (s2js.isUndefined(self.getInternalJsArray()[n])) {
            throw new scala.NoSuchElementException('An item with index ' + n + ' is not present.');
        }
        return self.getInternalJsArray()[n];
    };
scala.collection.Seq.prototype.update = function(n, newelem) {
var self = this;

        if (self.size() <= n) {
            throw new scala.NoSuchElementException('An item with index ' + n + ' is not present.');
        }
        self.getInternalJsArray()[n] = newelem;
    };
scala.collection.Seq.prototype.$length = function() {
var self = this;
return self.size();
};
scala.collection.Seq.prototype.remove = function(index) {
var self = this;

        if (index < 0 || self.size() <= index) {
            throw new scala.NoSuchElementException('An item with index ' + n + ' is not present.');
        }
        var removed = self.getInternalJsArray()[index];
        self.getInternalJsArray().splice(index, 1);
        return removed;
    };
scala.collection.Seq.prototype.prepend = function(x) {
var self = this;
self.getInternalJsArray().splice(0, 0, x);};
scala.collection.Seq.prototype.$minus$eq = function(x) {
var self = this;

        var index = self.getInternalJsArray().indexOf(x);
        if (index != -1) {
            self.getInternalJsArray().splice(index, 1);
        }
    };
scala.collection.Seq.prototype.indexWhere = function(p, from) {
var self = this;
if (typeof(from) === 'undefined') { from = 0; }
var i = from;
scala.util.control.Breaks.breakable(function() {
self.drop(from).foreach(function(x) {
if (p(x)) {
scala.util.control.Breaks.$break();
} else {
i = (i + 1);
}

});
i = -1;

});
return i;

};
scala.collection.Seq.prototype.diff = function(that) {
var self = this;

        return self.getInternalJsArray().filter(function(i) {
            return !(that.getInternalJsArray().indexOf(i) > -1);
        });
        };
scala.collection.Seq.prototype.__class__ = new s2js.Class('scala.collection.Seq', [scala.collection.Iterable]);
